src/sign/generic.rs

```diff
@@ -0,0 +1,512 @@
+                writeln!(w, "Algorithm: 8 (RSASHA256)")?;
+                k.into_dns(w)
+            }
+
+            Self::EcdsaP256Sha256(s) => {
+                writeln!(w, "Algorithm: 13 (ECDSAP256SHA256)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::EcdsaP384Sha384(s) => {
+                writeln!(w, "Algorithm: 14 (ECDSAP384SHA384)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::Ed25519(s) => {
+                writeln!(w, "Algorithm: 15 (ED25519)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+
+            Self::Ed448(s) => {
+                writeln!(w, "Algorithm: 16 (ED448)")?;
+                writeln!(w, "PrivateKey: {}", base64::encode_display(s))
+            }
+        }
+    }
+
+    /// Parse a key from the conventional DNS format.
+    ///
+    /// - For RSA, see RFC 5702, section 6.
+    /// - For ECDSA, see RFC 6605, section 6.
+    /// - For EdDSA, see RFC 8080, section 6.
+    pub fn from_dns(data: &str) -> Result<Self, DnsFormatError>
+    where
+        B: From<Vec<u8>>,
+    {
+        /// Parse private keys for most algorithms (except RSA).
+        fn parse_pkey<const N: usize>(
+            data: &str,
+        ) -> Result<[u8; N], DnsFormatError> {
+            // Extract the 'PrivateKey' field.
+            let (_, val, data) = parse_dns_pair(data)?
+                .filter(|&(k, _, _)| k == "PrivateKey")
+                .ok_or(DnsFormatError::Misformatted)?;
+
+            if !data.trim().is_empty() {
+                // There were more fields following.
+                return Err(DnsFormatError::Misformatted);
+            }
+
+            let buf: Vec<u8> = base64::decode(val)
+                .map_err(|_| DnsFormatError::Misformatted)?;
+            let buf = buf
+                .as_slice()
+                .try_into()
+                .map_err(|_| DnsFormatError::Misformatted)?;
+
+            Ok(buf)
+        }
+
+        // The first line should specify the key format.
+        let (_, _, data) = parse_dns_pair(data)?
+            .filter(|&(k, v, _)| (k, v) == ("Private-key-format", "v1.2"))
+            .ok_or(DnsFormatError::UnsupportedFormat)?;
+
+        // The second line should specify the algorithm.
+        let (_, val, data) = parse_dns_pair(data)?
+            .filter(|&(k, _, _)| k == "Algorithm")
+            .ok_or(DnsFormatError::Misformatted)?;
+
+        // Parse the algorithm.
+        let mut words = val.split_whitespace();
+        let code = words
+            .next()
+            .ok_or(DnsFormatError::Misformatted)?
+            .parse::<u8>()
+            .map_err(|_| DnsFormatError::Misformatted)?;
+        let name = words.next().ok_or(DnsFormatError::Misformatted)?;
+        if words.next().is_some() {
+            return Err(DnsFormatError::Misformatted);
+        }
+
+        match (code, name) {
```

Comment:
Yes, this is deliberately made a subset of all keys that we should validate -- validation related code belongs in the `validate` submodule.  I'm probably going to restructure this so that `sign` depends on `validate` and uses their public key type definitions.